homeassistant/components/husqvarna_automower/coordinator.py

```diff
@@ -82,14 +74,38 @@ async def _async_update_data(self) -> MowerDictionary:
             raise UpdateFailed(err) from err
         except AuthError as err:
             raise ConfigEntryAuthFailed(err) from err
-        self._async_add_remove_devices_and_entities(data)
         return data
 
+    @callback
+    def _on_data_update(self) -> None:
+        """Handle data updates and process dynamic entity management."""
+        if self.data is not None:
+            self._async_add_remove_devices()
+            for mower_id in self.data:
+                if self.data[mower_id].capabilities.stay_out_zones:
+                    self._async_add_remove_stay_out_zones()
+                if self.data[mower_id].capabilities.work_areas:
+                    self._async_add_remove_work_areas()
+
     @callback
     def handle_websocket_updates(self, ws_data: MowerDictionary) -> None:
         """Process websocket callbacks and write them to the DataUpdateCoordinator."""
+        self.hass.async_create_task(self._process_websocket_update(ws_data))
+
+    async def _process_websocket_update(self, ws_data: MowerDictionary) -> None:
+        """Handle incoming websocket update and update coordinator data."""
+        for data in ws_data.values():
+            existing_areas = data.work_areas or {}
+            for task in data.calendar.tasks:
+                work_area_id = task.work_area_id
+                if work_area_id is not None and work_area_id not in existing_areas:
+                    _LOGGER.debug(
+                        "New work area %s detected, refreshing data", work_area_id
+                    )
+                    await self.async_refresh()
```

Comment:
Could consider to use `async_schedule_refresh()` instead to avoid directly calling in case there is more requests within debouncing time.

Reply:
I can only offer `async_request_refresh`